internal/providers/vault/provider.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+	"sync"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+func Register() {
+	p := &provider{
+		vaultClients: make(map[string]*vaultclient.Client),
+	}
+	combined.RegisterSecretsProvider(p.appRoleProvider)
+	combined.RegisterSecretsProvider(p.certificateProvider)
+}
+
+type provider struct {
+	mtx          sync.Mutex
+	vaultClients map[string]*vaultclient.Client
+}
+
+func (p *provider) Login(ctx context.Context, caCfg *vault.VaultProvider, audience string) (*vaultclient.Client, error) {
+	vKey := vaultIdentifier{
```

Comment:
Nit: `key` would have been unambiguous enough. :-P

Reply:
Indeed! This file is copy/paste though, but Git doesn't seem to have recognised that.